services/placement/algo/non_sharded.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package algo
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/services/placement"
+)
+
+var errAddReplicaOnNonSharded = errors.New("could not add replica on nonsharded placement")
+
+type nonShardedAlgo struct{}
+
+func newNonShardedAlgo() placement.Algorithm {
+	return nonShardedAlgo{}
+}
+
+func (a nonShardedAlgo) InitialPlacement(
+	instances []services.PlacementInstance,
+	shards []uint32,
+) (services.ServicePlacement, error) {
+	return placement.NewPlacement(cloneInstances(instances), shards, 0), nil
+}
+
+func (a nonShardedAlgo) AddReplica(p services.ServicePlacement) (services.ServicePlacement, error) {
+	return nil, errAddReplicaOnNonSharded
+}
+
+func (a nonShardedAlgo) RemoveInstance(p services.ServicePlacement, instanceID string) (services.ServicePlacement, error) {
+	instances := p.Instances()
+	for i, instance := range instances {
+		if instance.ID() == instanceID {
+			instances = append(instances[:i], instances[i+1:]...)
```

Comment:
can perhaps swap it with the last element and shrink the slice size to avoid a copy?